### PR TITLE
Adapt bind_socket to make it usable with multiple processes

### LINKED ISSUE #428: UNIX domain socket raises 503 when workers are used
Is it Ok when a combination of `--workers ` and `--uds` raises 503 error via nginx? For example:
```
--workers 4 --uds /var/run/uvicorn-test.sock
```

it works well when `--workers` is omitted (1 worker). Please note that `gunicorn` works well via Unix socket and multiple workers.

Thanks!

### PR DESCRIPTION
**Pull request by euri10**
This makes it possible to use reload and workers flag with uds and fd.

fixes https://github.com/encode/uvicorn/issues/428 https://github.com/encode/uvicorn/issues/368 https://github.com/encode/uvicorn/issues/586 #722

### CODE REVIEW

**Comment by euri10** [comment]
still WIP I noticed the uvicorn logger message needs to be adapted

**Comment by euri10** [comment]
ready now @Kludex  !

File: uvicorn/config.py [hunk 1]
```diff
@@ -366,37 +366,75 @@ def setup_event_loop(self):
...
+            )
+            protocol_name = "https" if self.is_ssl else "http"
+            logger.info(
+                message,
+                protocol_name,
+                self.host,
+                self.port,
+                extra={"color_message": color_message},
+            )
         sock.set_inheritable(True)
```
Comment (Kludex) [inline uvicorn/config.py:454]: Does this impact the socket from `socket.fromfd`? https://docs.python.org/3/library/socket.html#socket.fromfd

image

File: uvicorn/config.py [hunk 2]
```diff
@@ -366,37 +366,75 @@ def setup_event_loop(self):
...
-            logger.error(exc)
-            sys.exit(1)
+            message = f"Uvicorn running on {addr_format} (Press CTRL+C to quit)"
+            color_message = (
+                "Uvicorn running on "
+                + click.style(addr_format, bold=True)
+                + " (Press CTRL+C to quit)"
+            )
+            protocol_name = "https" if self.is_ssl else "http"
+            logger.info(
```
Comment (Kludex) [inline uvicorn/config.py]: Those `logger.info` take a lot of lines now, can we have a single one at the end that accepts `logger_args` and the `color_message`?

This is just a small point not related to the logic, so you can close it if you think this is not necessary / overwork. :kissing:

File: tests/test_config.py [hunk 3]
```diff
@@ -320,3 +321,25 @@ def test_ws_max_size():
...
+        (True, 1),
+        (False, 2),
+    ],
+    ids=["--reload=True --workers=1", "--reload=False --workers=2"],
+)
+@pytest.mark.skipif(sys.platform == "win32", reason="require unix-like system")
+def test_bind_unix_socket_works_with_reload_or_workers(tmp_path, reload, workers):
+    uds_file = tmp_path / "uvicorn.sock"
+    config = Config(
+        app=asgi_app, uds=uds_file.as_posix(), reload=reload, workers=workers
```
Comment (Kludex) [inline tests/test_config.py:372]: Is there a reason for not adding any test for `fd`?

**Review by Kludex — COMMENTED** [review]
I did my best here. It's a short PR, but I feel like I'm not qualified to approve it yet. I'm going to study some details before having the confidence to do so.

But I hope my current comments make sense.

File: uvicorn/config.py (hunk 1 above)
Comment (euri10) [reply uvicorn/config.py:454]: this `sock.set_inheritable(True)` applies to all 3 cases so yes. it serves when we have multiple process and 1 socket to listen on iirc

File: uvicorn/config.py (hunk 2 above)
Comment (euri10) [reply uvicorn/config.py]: yes that's a good idea, done

File: tests/test_config.py (hunk 3 above)
Comment (euri10) [reply tests/test_config.py:372]: no, I added one

**Comment by euri10** [comment]
ready for another review @Kludex 
let me know if something not clear I can easily expain

File: uvicorn/config.py (hunk 1 above)
Comment (Kludex) [reply uvicorn/config.py:454]: Ah, I understand what confused me now. Thanks.

File: uvicorn/main.py [hunk 4]
```diff
@@ -391,6 +392,8 @@ def run(app: typing.Union[ASGIApplication, str], **kwargs: typing.Any) -> None:
         Multiprocess(config, target=server.run, sockets=[sock]).run()
     else:
         server.run()
+        if config.uds:
+            os.remove(config.uds)
```
Comment (Kludex) [inline uvicorn/main.py]: Maybe I'm mistaken, but can't we have a single one of those outside the conditional?

```suggestion
    if config.uds:
        os.remove(config.uds)
```

Comment (Kludex) [reply uvicorn/main.py]: I guess we can, but it's not right, because we bind inside them. :thinking: Right?

Comment (euri10) [reply uvicorn/main.py]: it looks simpler yes, I removed occurrences in the basereload and multiprocess shutdown method and put it outside

**Review by Kludex — APPROVED** [review]
I'm confident enough now.

But maybe we should wait for the `mypy` merge first? Anyway, your call.

**Comment by euri10** [comment]
> But maybe we should wait for the `mypy` merge first? Anyway, your call.

the one on config.py ?

**Comment by Kludex** [comment]
Yeah

**Comment by euri10** [comment]
> Yeah

that's a good idea, he put a lot of efforts into it and it would be cruel to have to add that part on top of it ;)

File: uvicorn/config.py [hunk 5]
```diff
@@ -395,37 +395,63 @@ def setup_event_loop(self) -> None:
             loop_setup()
 
     def bind_socket(self) -> socket.socket:
-        family = socket.AF_INET
-        addr_format = "%s://%s:%d"
+        logger_args: List[Union[str, int]]
```
Comment (euri10) [inline uvicorn/config.py:398]: just needed to add this post-merge with master @Kludex 
the rest stayed the same

Comment (euri10) [reply uvicorn/config.py:398]: So  I removed the hold button and feel free to merge if still ok with this !

Comment (Kludex) [reply uvicorn/config.py:398]: I find it so weird that the approval is not removed after a commit is pushed. I don't understand the reason behind it.

Anyway, LGTM! :)

**Pull request merged**